### PR TITLE
Add 'with' option to Include tag

### PR DESCRIPTION
**Pull request by hishamco**
This a second part of the issue #30

### CODE REVIEW

**Comment by hishamco** [comment]
@sebastienros this is straightforward, so after I got your comments about the PR #32, I can rebase to include this change

File: Fluid/Ast/IncludeStatement.cs [hunk 1]
```diff
@@ -22,13 +22,16 @@ public class IncludeStatement : Statement
...
+            Value = value;
         }
 
         public Expression Path { get; }
 
+        public Expression Value { get; }
+
         public override async Task<Completion> WriteToAsync(TextWriter writer, TextEncoder encoder, TemplateContext context)
         {
             var relativePath = (await Path.EvaluateAsync(context)).ToStringValue();
```
Comment (sebastienros) [inline Fluid/Ast/IncludeStatement.cs]: Maybe save the value at this point, and not call `Path.EvaluateAsync` again ?

File: Fluid/Ast/IncludeStatement.cs [hunk 2]
```diff
@@ -22,13 +22,16 @@ public class IncludeStatement : Statement
...
-        public IncludeStatement(Expression path)
+        public IncludeStatement(Expression path, Expression value = null)
         {
             Path = path;
+            Value = value;
         }
 
         public Expression Path { get; }
 
+        public Expression Value { get; }
```
Comment (sebastienros) [inline Fluid/Ast/IncludeStatement.cs]: Should it be named `With` to make it obvious what kind of value it is?

File: Fluid/DefaultFluidParser.cs [hunk 3]
```diff
@@ -527,7 +527,13 @@ public static DecrementStatement BuildDecrementStatement(ParseTreeNode tag)
         public static IncludeStatement BuildIncludeStatement(ParseTreeNode tag)
         {
             var pathExpression = BuildTermExpression(tag.ChildNodes[0]);
-            return new IncludeStatement(pathExpression);
+            Expression valueExpression = null;
```
Comment (sebastienros) [inline Fluid/DefaultFluidParser.cs]: `withExpression` ?

File: Fluid/Ast/IncludeStatement.cs (hunk 1 above)
Comment (hishamco) [reply Fluid/Ast/IncludeStatement.cs]: It's similar to `Path`?!!

File: Fluid/Ast/IncludeStatement.cs [hunk 4]
```diff
@@ -45,6 +48,14 @@ public IncludeStatement(Expression path)
                 throw new FileNotFoundException(relativePath);
             }
 
+            if (Value != null)
+            {
+                var identifier = (await Path.EvaluateAsync(context)).ToStringValue();
+
+                await new AssignStatement(identifier, Value).
+                    WriteToAsync(writer, encoder, context);
```
Comment (sebastienros) [inline Fluid/Ast/IncludeStatement.cs]: This should be evaluated on `childScope` instead of `context`

File: Fluid/Ast/IncludeStatement.cs (hunk 1 above)
Comment (hishamco) [reply Fluid/Ast/IncludeStatement.cs]: But we need the value of the context!!

Comment (sebastienros) [reply Fluid/Ast/IncludeStatement.cs]: If I change the code and the tests pass, will you trust me ?

Comment (hishamco) [reply Fluid/Ast/IncludeStatement.cs]: Be free to change the code if you have a better solution

> will you trust me ?

I'm always trust you 😄

**Comment by hishamco** [comment]
I see, because I surprised when you told me grap the result of `context.EnterChildScope()` and pass it to the `WriteToAsync` , coz the scope can't be converted to context 😄

**Comment by sebastienros** [comment]
Do you want me to merge one PR and you'll update the other one ?

**Comment by hishamco** [comment]
Yep, that's what I wrote in the other PR, do it please ..

File: Fluid/Ast/IncludeStatement.cs [hunk 5]
```diff
@@ -45,6 +48,21 @@ public IncludeStatement(Expression path)
...
+                var identifier = System.IO.Path.GetFileNameWithoutExtension(relativePath);
+                context.EnterChildScope();
+
+                try
+                {
+                    await new AssignStatement(identifier, With).WriteToAsync(writer, encoder, context);
+                }
+                catch
+                {
+                    context.ReleaseScope();
```
Comment (sebastienros) [inline Fluid/Ast/IncludeStatement.cs]: I think that's wrong, and the tests should have shown it.

**Comment by sebastienros** [comment]
Please add a test that ensures that a variable is not set in the template that calls the include tag. For `with` and the inline parameters. This should have failed, the usage of EnterChildScope was wrong.

**Comment by hishamco** [comment]
@sebastienros as you said before, there's an issue with `EnterChildScope`

**Comment by hishamco** [comment]
👍

**Pull request merged**